rasterio/_io.pyx

```diff
@@ -214,18 +217,31 @@ cdef class DatasetReaderBase(DatasetBase):
             dtype = check_dtypes.pop()
 
         # Get the natural shape of the read window, boundless or not.
+        # The window can have float values. In this case, we round up
+        # when computing the shape.
+
+        # Stub the win_shape.
         win_shape = (len(indexes),)
+
         if window:
+
             if boundless:
                 win_shape += (
-                        window[0][1]-window[0][0], window[1][1]-window[1][0])
+                    int(round(window[0][1] - window[0][0], 6)),
+                    int(round(window[1][1] - window[1][0], 6)))
```

Comment:
These kind of subtractions and the potential loss of float precision are why I redid the `Window` class. TODO: change this code.